src/ray/id.h

```diff
@@ -2,44 +2,128 @@
 #define RAY_ID_H_
 
 #include <inttypes.h>
+#include <limits.h>
 
+#include <chrono>
 #include <cstring>
+#include <mutex>
+#include <random>
 #include <string>
 
 #include "plasma/common.h"
 #include "ray/constants.h"
+#include "ray/util/logging.h"
 #include "ray/util/visibility.h"
 
 namespace ray {
 
-class RAY_EXPORT UniqueID {
+class DriverID;
+class UniqueID;
+
+// Declaration.
+std::mt19937 RandomlySeededMersenneTwister();
+uint64_t MurmurHash64A(const void *key, int len, unsigned int seed);
+
+// Change the compiler alignment to 1 byte (default is 8).
+#pragma pack(push, 1)
+
+template <typename T>
+class BaseID {
  public:
-  UniqueID();
-  UniqueID(const plasma::UniqueID &from);
-  static UniqueID from_random();
-  static UniqueID from_binary(const std::string &binary);
-  static const UniqueID &nil();
+  BaseID();
```

Comment:
There must not be virtual functions. If there is, there will be virtual table pointer which is suitable for current structure.

Reply:
But if the dtor is not virtual, it may cause memory leak(or other more fatal things) although it hasn't happened yet.

Considering there is anther developer may add `XXXId` extending `BaseId`, and `XXXId` has a dtor which does some clean up:
```C++
class XXXId : public BaseId {
 public:
  ~XXXId() {
    // do some clean up, maybe free memory.
  }
};
```
And once some one uses the class as polymorphic, the dtor of `XXXId` will never be called because the dtor of `BaseId` is non-virtual.

Reply:
We don't need the destructor to clean the memory. To avoid other developer to add the virtual functions, there is `static_assert` to make compiler to help us to check.